deployment/ccip/changeset/prerequisites.go

```diff
@@ -0,0 +1,111 @@
+package changeset
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/ccip-owner-contracts/pkg/proposal/timelock"
+	chain_selectors "github.com/smartcontractkit/chain-selectors"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	ccipdeployment "github.com/smartcontractkit/chainlink/deployment/ccip"
+)
+
+var (
+	_                        deployment.ChangeSet[PrerequisiteConfig] = InitializePrerequisites
+	mapPrerequisiteContracts                                          = map[deployment.ContractType]struct{}{
+		ccipdeployment.LinkToken:          {},
+		ccipdeployment.WETH9:              {},
+		ccipdeployment.TokenAdminRegistry: {},
+		ccipdeployment.RegistryModule:     {},
+		ccipdeployment.Router:             {},
+	}
+)
+
+// InitializePrerequisites loads the existing contracts into the address book.
+// This is required for contracts which can be reused from previous versions of CCIP
+// If PrerequisiteConfig.Deploy is true, it will deploy the prerequisite contracts except Router
+// Router is deployed as part of DeployChainContracts Changeset due to its dependency on RMNProxy
+func InitializePrerequisites(env deployment.Environment, cfg PrerequisiteConfig) (deployment.ChangesetOutput, error) {
+	err := cfg.Validate()
+	if err != nil {
+		return deployment.ChangesetOutput{}, errors.Wrapf(deployment.ErrInvalidConfig, "%v", err)
+	}
+	ab := deployment.NewMemoryAddressBook()
+	if cfg.Deploy {
+		err = ccipdeployment.DeployPrerequisiteContracts(env, ab, cfg.ChainSelectors)
+		if err != nil {
+			env.Logger.Errorw("Failed to deploy prerequisite contracts", "err", err, "addressBook", ab)
+			return deployment.ChangesetOutput{}, fmt.Errorf("failed to deploy prerequisite contracts: %w", err)
+		}
+		return deployment.ChangesetOutput{
+			Proposals:   []timelock.MCMSWithTimelockProposal{},
+			AddressBook: ab,
+			JobSpecs:    nil,
+		}, nil
+	}
+	for _, sel := range cfg.ChainSelectors {
+		for _, ec := range cfg.ExistingContracts {
+			err = ab.Save(sel, ec.Address.String(), ec.TypeAndVersion)
+			if err != nil {
+				env.Logger.Errorw("Failed to deploy prerequisite contracts", "err", err, "addressBook", ab)
+				return deployment.ChangesetOutput{}, fmt.Errorf("failed to save existing contract: %w", err)
+			}
+		}
+	}
+
+	return deployment.ChangesetOutput{
+		Proposals:   []timelock.MCMSWithTimelockProposal{},
+		AddressBook: ab,
+		JobSpecs:    nil,
+	}, nil
+}
+
+type ContractConfig struct {
+	Address        common.Address
+	TypeAndVersion deployment.TypeAndVersion
+}
+
+type PrerequisiteConfig struct {
+	ChainSelectors []uint64
+	Deploy         bool
+	// if Deploy is false, ExistingContracts must be set
```

Comment:
I think cleaner to have a separate changeset like SaveExistingContracts which takes in an arbitrary number of []Contract{selector, address, typeAndVersion} and just saves them to an address book and returns it. That would be more generally reusable and avoid the conditional